build: Replace `which` command with `command -v`

hebasto opened the pull request:
On some systems the `which` command can emit messages into stderr. For example, for `debianutils 5.5-1` package in Debian Sid:
```
# which cat  
/usr/bin/which: this version of `which' is deprecated; use `command -v' in scripts instead.
/bin/cat
```

Although such messages are harmless, they could distract developers needlessly (see bitcoin/bitcoin#24056).

Fixes bitcoin/bitcoin#24056.

hebasto committed:
build: Replace `which` command with `command -v`

This change made in a way that is compatible with GNU Make versions
older than 4.3.

kristapsk commented:
Can't we switch to using `command -v` instead? Or there are platforms where `which` is available but shell does not have `command` built-in?

hebasto commented:
> Can't we switch to using `command -v` instead? Or there are platforms where `which` is available but shell does not have `command` built-in?

https://github.com/bitcoin/bitcoin/issues/24056#issuecomment-1021596825

laanwj commented:
Thanks for checking about `command -v` avoiding a next round of hassle.

Tested ACK https://github.com/bitcoin/bitcoin/pull/24156/commits/1c45a9a900f35cf400392f926af0e08ae42d709f
(tested that the message didn't appear, will do a full GUIX build to make sure the windows build works)

katesalazar commented:
ACK 1c45a9a900f35cf400392f926af0e08ae42d709f

katesalazar commented on depends/hosts/mingw32.mk:
```diff
@@ -1,4 +1,4 @@
-ifneq ($(shell which $(host)-g++-posix),)
+ifneq ($(shell which $(host)-g++-posix 2> /dev/null),)
```
(nit) 2>/dev/null (no space) is more elegant IMO

fanquake commented:
The main problem this is fixing isn't

> /usr/bin/which: this version of `which' is deprecated;

it's the output from #24056 which was caused by #22093.

I mentioned [here](https://github.com/bitcoin/bitcoin/issues/24056#issuecomment-1022789731), we already use `command -v`, so I'm not sure why we can't use it in this case.

katesalazar commented:
fanquake wrote:

> I mentioned [here](https://github.com/bitcoin/bitcoin/issues/24056#issuecomment-1022789731), we already use `command -v`, so I'm not sure why we can't use it in this case.

hebasto [wrote][0]:

> Will add some more tests and then submit a pr.

I understand you can and is coming in a different PR.

[0]: https://github.com/bitcoin/bitcoin/issues/24056#issuecomment-1021312655

hebasto commented:
@fanquake 
> The main problem this is fixing isn't
> 
> > /usr/bin/which: this version of `which' is deprecated;
> 
> it's the output from #24056 which was caused by #22093.

Both output from #24056 and another example in the OP are generally the same -- the stderr output.

> I mentioned [here](https://github.com/bitcoin/bitcoin/issues/24056#issuecomment-1022789731), we already use `command -v`, so I'm not sure why we can't use it in this case.

`grep`ping gives the only place: https://github.com/bitcoin/bitcoin/blob/bd482b3ffebc68130f8a18dabf08ed1aff7ea159/depends/hosts/darwin.mk#L28-L36

and comment added in 880660acfa547558f6ef5adff6768de95e53af6e by @dongcarl (#19683) seems say the same as https://github.com/bitcoin/bitcoin/issues/24056#issuecomment-1021596825.

luke-jr commented:
Doesn't #24131 resolve this without the ugly stderr loss? (Would be bad to silence an actually-relevant warning/error in the future..)

prusnak commented:
> Doesn't #24131 resolve this without the ugly stderr loss? (Would be bad to silence an actually-relevant warning/error in the future..)

I don't think it does, because on older system you will not have the `...-g++-posix` binary.

hebasto commented:
Updated 1c45a9a900f35cf400392f926af0e08ae42d709f -> 148b33cf72033eced8a701d127e21dfe8a816ce3 ([pr24156.01](https://github.com/hebasto/bitcoin/commits/pr24156.01) -> [pr24156.02](https://github.com/hebasto/bitcoin/commits/pr24156.02)).

Addressed:

- @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/24156#issuecomment-1022791254)
- @luke-jr's [comment](https://github.com/bitcoin/bitcoin/pull/24156#issuecomment-1025319815)

dongcarl commented:
Code Review ACK 148b33cf72033eced8a701d127e21dfe8a816ce3

Somewhat frustrating upstream bug, but it is what it is 🤷

laanwj commented:
Code review ACK 148b33cf72033eced8a701d127e21dfe8a816ce3

> I don't think it does, because on older system you still will not have the ...-g++-posix binary.

True. But the problem here isn't so much older systems. It's that when doing a guix build, depends is used to download the packages. This is done outside the build sandbox (because no network connectivity is available there). There, no C++ compiler needs to be installed at all, let alone a mingw one.
